In [469]:
import pandas as pd
import numpy as np
import mlxtend
import ast
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
from mlxtend.preprocessing import TransactionEncoder
df = pd.read_csv("filtered_ingred.csv")

In [470]:
df.head()

Unnamed: 0                                      Filtered_Text
0           0  ['butter', 'apples', 'firm', 'thick', 'sugar',...
1           1  ['apples', 'butter', 'flour', 'sugar', 'water'...
2           2           ['apples', 'water', 'sugar', 'cinnamon']
3           3  ['apples', 'sugar', 'flour', 'cinnamon', 'wate...
4           4  ['apples', 'juice', 'water', 'sugar', 'cornsta...

In [471]:
#result=ast.literal_eval(df['Filtered_Text'])

In [472]:
#df['Filtered_Text']=df.Filtered_Text.str.split(',').values.tolist()
df=df['Filtered_Text'].values


In [473]:
initial_df=df.tolist()
result=np.array(initial_df,dtype=object)

In [474]:
result

array(["['butter', 'apples', 'firm', 'thick', 'sugar', 'cranberries', 'sheet', 'puff', 'pastry', 'package', 'directions', 'egg', 'cream']",
       "['apples', 'butter', 'flour', 'sugar', 'water', 'inch', 'double-crust', 'pie', 'pastry']",
       "['apples', 'water', 'sugar', 'cinnamon']", ...,
       "['chile', 'peppers', 'onion', 'dates', 'basil', 'leaves', 'parsley', 'roma', 'plum', 'tomato', 'beef', 'bouillon', 'powder', 'cloves', 'oil', 'salt', 'squeeze', 'lemon', 'juice']",
       "['dates', 'butter', 'boiling', 'water', 'sugar', 'egg', 'flour', 'baking', 'soda', 'powder', 'salt', 'walnuts']",
       "['dates', 'cashews', 'almonds', 'high-quality', 'cocoa', 'powder', 'oil', 'coconut', 'vanilla', 'extract', 'cold', 'espresso', 'water', 'kosher', 'salt', 'cayenne', 'pepper']"],
      dtype=object)

In [475]:
for i in range(len(result)):
    result[i]=ast.literal_eval(result[i])

In [476]:
result

array([list(['butter', 'apples', 'firm', 'thick', 'sugar', 'cranberries', 'sheet', 'puff', 'pastry', 'package', 'directions', 'egg', 'cream']),
       list(['apples', 'butter', 'flour', 'sugar', 'water', 'inch', 'double-crust', 'pie', 'pastry']),
       list(['apples', 'water', 'sugar', 'cinnamon']), ...,
       list(['chile', 'peppers', 'onion', 'dates', 'basil', 'leaves', 'parsley', 'roma', 'plum', 'tomato', 'beef', 'bouillon', 'powder', 'cloves', 'oil', 'salt', 'squeeze', 'lemon', 'juice']),
       list(['dates', 'butter', 'boiling', 'water', 'sugar', 'egg', 'flour', 'baking', 'soda', 'powder', 'salt', 'walnuts']),
       list(['dates', 'cashews', 'almonds', 'high-quality', 'cocoa', 'powder', 'oil', 'coconut', 'vanilla', 'extract', 'cold', 'espresso', 'water', 'kosher', 'salt', 'cayenne', 'pepper'])],
      dtype=object)

In [477]:
from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()
transaction_matrix = te.fit_transform(result, sparse=True)
transaction_matrix

<1090x690 sparse matrix of type '<class 'numpy.bool_'>'
	with 11294 stored elements in Compressed Sparse Row format>

In [478]:
transaction_df = pd.DataFrame.sparse.from_spmatrix(transaction_matrix, columns = te.columns_)
transaction_df.head()

%  Conference  achiote  adobo  agave  ale  alfalfa  allspice  almond  \
0  0           0        0      0      0    0        0         0       0   
1  0           0        0      0      0    0        0         0       0   
2  0           0        0      0      0    0        0         0       0   
3  0           0        0      0      0    0        0         0       0   
4  0           0        0      0      0    0        0         0       0   

   almondmilk  ...  yolks  zest  zucchini  ®  ¼  ¾  ⅓  ⅔  ⅛  ⅞  
0           0  ...      0     0         0  0  0  0  0  0  0  0  
1           0  ...      0     0         0  0  0  0  0  0  0  0  
2           0  ...      0     0         0  0  0  0  0  0  0  0  
3           0  ...      0     0         0  0  0  0  0  0  0  0  
4           0  ...      0     0         0  0  0  0  0  0  0  0  

[5 rows x 690 columns]

In [479]:
mlxtend.__version__

'0.22.0'

In [480]:
#feed into apriori
x = apriori(transaction_df, min_support = 0.02, use_colnames=True,low_memory=True)

In [481]:
x.sort_values(by = 'support', ascending = False)

support                                      itemsets
91    0.667890                                       (sugar)
83    0.422018                                        (salt)
17    0.355046                                      (butter)
43    0.310092                                       (flour)
182   0.300917                               (butter, sugar)
...        ...                                           ...
1435  0.020183     (vanilla, butter, baking, flour, extract)
1434  0.020183       (butter, baking, flour, sugar, extract)
1807  0.020183  (powder, butter, eggs, flour, salt, extract)
187   0.020183                               (cake, package)
556   0.020183                      (vanilla, butter, cream)

[1934 rows x 2 columns]

In [482]:
rules_df = association_rules(x, metric = 'lift', min_threshold=1)
rules_df.sort_values(by='lift',ascending = False)

antecedents      consequents  antecedent support  \
233            (garlic)          (clove)            0.027523   
232             (clove)         (garlic)            0.021101   
80           (skinless)       (boneless)            0.020183   
81           (boneless)       (skinless)            0.032110   
143               (mix)           (cake)            0.043119   
...                 ...              ...                 ...   
6924  (cinnamon, sugar)  (extract, milk)            0.242202   
423             (sugar)          (lemon)            0.667890   
422             (lemon)          (sugar)            0.126606   
3175           (powder)   (pepper, salt)            0.194495   
3170     (pepper, salt)         (powder)            0.145872   

      consequent support   support  confidence       lift  leverage  \
233             0.021101  0.021101    0.766667  36.333333  0.020520   
232             0.027523  0.021101    1.000000  36.333333  0.020520   
80              0.032110  0.020183    1.000000  31.142857  0.019535   
81              0.020183  0.020183    0.628571  31.142857  0.019535   
143             0.024771  0.022018    0.510638  20.614657  0.020950   
...                  ...       ...         ...        ...       ...   
6924            0.082569  0.020183    0.083333   1.009259  0.000185   
423             0.126606  0.085321    0.127747   1.009018  0.000763   
422             0.667890  0.085321    0.673913   1.009018  0.000763   
3175            0.145872  0.028440    0.146226   1.002433  0.000069   
3170            0.194495  0.028440    0.194969   1.002433  0.000069   

      conviction  zhangs_metric  
233     4.195282       1.000000  
232          inf       0.993440  
80           inf       0.987828  
81      2.637968       1.000000  
143     1.992860       0.994367  
...          ...            ...  
6924    1.000834       0.012107  
423     1.001309       0.026911  
422     1.018471       0.010233  
3175    1.000416       0.003013  
3170    1.000588       0.002841  

[31632 rows x 10 columns]

In [483]:
rules_df['antecedents'] = rules_df['antecedents'].apply(lambda x:list(x)).copy()
rules_df['consequents'] = rules_df['consequents'].apply(lambda x:list(x)).copy()

In [489]:
import numpy as np
# Input basket
mybasket = ['salt', 'apple', 'sugar','butter']

#metric
metric = 'lift'

#COMPLETE THIS FUNCTION
def product_recs(basket, rule_df, metric):
    
    # Randomly select an item from the basket
    random_item = np.random.choice(basket, 1)[0]
    print(random_item)
    
    # Find rules where the item is in the antecedent
    rule_filter = rule_df['antecedents'].apply(lambda x:x[0]) == random_item
    
    # Filter the dataframe using rule_filter and sort by the selected metric
    filtered_df = rule_df[rule_filter].sort_values(by = metric, ascending = False)
    
    # Randomly return one of the top 20 items from the filtered dataframe
    reco = filtered_df.head(20).sample()['consequents']
    
    return reco

product_recs(mybasket, rules_df, metric )

butter


1466    [pie]
Name: consequents, dtype: object

If user has salt,apple,sugar and butter, it is recommended to shop pie and cook recipe number 1466